### ISSUE TITLE
Spyglass-sample issues multiple mention would not trigger after remove space from wordBreakChars

### ISSUE THREAD

**Issue by elaychan** [linkedin/Spyglass#92]
**Issue:**
During testing with the spyglass-sample , I have encounter a "funny" situation in GridMentions Demo.

If I want to allow user to type space when searching user "Joan Wheeler", from my understand I will have to override the default `wordBreakChars` from the `WordTokenizerConfig.Builder` by `setWordBreakChars(",." + System.getProperty("line.separator"))` in order to remove " " from the breaking keyword. 
But if I do so, user can only add tag one user, unless user type either "," / "." / "\n" keyword to enable to trigger the tag mention mechanism again. 

I have tried to change the number of keywords, but it does not help. The following is the code I used for the tokenizer, everything is the same as the sample app for GridMentions. 
**Code:**

```
private final WordTokenizerConfig tokenizerConfig = new WordTokenizerConfig
            .Builder()
            .setWordBreakChars("." + System.getProperty("line.separator"))
            .setExplicitChars("@")
            // Set the number of word that can be detected. (Should have able to detect 2 words)
            .setMaxNumKeywords(2)
            // Set the threshold to 150 in order to have the mention mechanism triggered only by "@"
            .setThreshold(150)
            .build();
```
I have read this [closed issue #16](https://github.com/linkedin/Spyglass/issues/16), setting low threshold ( 0, 1, 2) would help trigger the mechanism again, but it will also generate more problem such as continuously trigger api to search for user.

**Comment by elaychan** [comment]
Apparently, library is not allow to call `setThreshold`, and `setMaxNumKeywords` at the same time.
If I removed `setThreshold`, and add a contain check in `onQueryReceived()` everything would work as I wanted. 

**Code:**
```
    private val tokenizerConfig = WordTokenizerConfig.Builder()
            .setWordBreakChars(" ." + System.getProperty("line.separator"))
            .setExplicitChars(Constants.TAG_TOKENIZER)
            .setMaxNumKeywords(10)
            .build()

override fun onQueryReceived(queryToken: QueryToken): List<String> {
        val buckets = listOf(BUCKET)
        if (queryToken.tokenString.first() == '@') {
            // do something
        }
        return buckets
    }
```
The above code would have achieved the following conditions:

- Use @ symbol to trigger mention mechanism 
- Allow user to type space during query searching
- Prevent user trigger mention mechanism when typing.

**Comment by nhibner** [comment]
> Apparently, library is not allow to call `setThreshold`, and `setMaxNumKeywords` at the same time.

I am not sure what you mean by this. Apps can definitely set different values for the threshold vs the maximum number of keywords. The threshold is the number of characters needed in a word to trigger an implicit mention. The maximum number of keywords is the greatest number of words that will be included in a query token.

> If I removed `setThreshold`, and add a contain check in `onQueryReceived()` everything would work as I wanted.

Note that removing `setThreshold` does not mean there is no threshold. It means that you would be using the default threshold value (which is 4).

As a side note, here are a few notes on the example code you posted above:

1. The WordTokenizerConfig can be simplified to this (note that the word-break characters and explicit characters you are setting are the same as the defaults):
```
private val tokenizerConfig = WordTokenizerConfig.Builder()
            .setMaxNumKeywords(10)
            .build()
```

2. I would suggest using `queryToken.isExplicit()` rather than `queryToken.tokenString.first() == '@'`

**Comment by elaychan** [comment]
I am sorry for the late reply, very busy lately in school. 

When I am playing around the library, I was going to set the following conditions:
- Use only the @ symbol to trigger mention mechanism
- Allow user to type space during query searching (eg. "John Doe is Awesome") 
- Prevent user trigger mention mechanism when typing.

I was having issues setting the correct Threshold, the correct number of keywords. As I have mentioned in the first post, I have read [Closed issue #16](https://github.com/linkedin/Spyglass/issues/16) which appears to have the same issue as I do. 

My original idea was only to allow users to tag other users only if explicitChars "@" has been enter by the user input. Without knowing I have to handle in which case I am allowing the app to trigger the tagging mechanism, I was assuming that setting a very high threshold would stop the issue of the tag suggestion keep popping up without the "@" keyword.

My assumption was correct, but then I have run into another issue, was allowing users to type space in between user tag searching( e.g space in between John Doe), with my assumption, typing space would result in the queryToken to change from "@John " to "D" after typing space and D. And setting umber of keyword does not seem to solve the problem.

Later I have found that, if I added in a contain explicit chars check, all issues will be resolved. Therefore I was thinking setting threshold and setting the number of keywords does not work together.

Which of course later I found out that it was because I am supposed to have my own logic to handle all these conditions myself.

But I still have one question, if we have explicit char which should trigger the tagging mechanism, why do we need to set number of thresholds?

**Comment by elaychan** [comment]
```
// Modified tokenizerConfig 
private final WordTokenizerConfig tokenizerConfig = new WordTokenizerConfig
            .Builder()
            .setWordBreakChars("." + System.getProperty("line.separator"))
            .setExplicitChars("@")
            // Set the number of word that can be detected.
            .setMaxNumKeywords(2)
            // Set the threshold to 150 in order to have the mention mechanism triggered only by "@"
            .setThreshold(150)
            .build();
// The original onQueryReceived function 
override fun onQueryReceived(queryToken: QueryToken): List<String> {
        List<String> buckets = Collections.singletonList(BUCKET);
        List<Person> suggestions = people.getSuggestions(queryToken);
        SuggestionsResult result = new SuggestionsResult(queryToken, suggestions);
        // Have suggestions, now call the listener (which is this activity)
        onReceiveSuggestionsResult(result, BUCKET);
        return buckets;
    }
```
Would behave like this:
https://imgur.com/a/MndamIz

And then I realized I have to handle the logic by myself:
```
// Modified tokenizerConfig 
private final WordTokenizerConfig tokenizerConfig = new WordTokenizerConfig
            .Builder()
            // Set the number of word that can be detected.
            .setMaxNumKeywords(2)
            // Set the threshold to 150 in order to have the mention mechanism triggered only by "@"
            .setThreshold(150)
            .build();
// The original onQueryReceived function 
override fun onQueryReceived(queryToken: QueryToken): List<String> {
        List<String> buckets = Collections.singletonList(BUCKET);
        List<Person> suggestions = people.getSuggestions(queryToken);
        SuggestionsResult result = new SuggestionsResult(queryToken, suggestions);
        // Have suggestions, now call the listener (which is this activity)
        onReceiveSuggestionsResult(result, BUCKET);
        return buckets;
    }
``` 
Which will have my desired behaviour: 
https://imgur.com/a/3tSabcR
